src/main/java/io/gravitee/policy/trafficshadowing/TrafficShadowingPolicy.java

```diff
@@ -15,30 +15,45 @@
  */
 package io.gravitee.policy.trafficshadowing;
 
-import io.gravitee.gateway.api.ExecutionContext;
-import io.gravitee.gateway.api.Invoker;
-import io.gravitee.policy.api.PolicyChain;
-import io.gravitee.policy.api.annotations.OnRequest;
+import io.gravitee.gateway.reactive.api.context.InternalContextAttributes;
+import io.gravitee.gateway.reactive.api.context.http.HttpPlainExecutionContext;
+import io.gravitee.gateway.reactive.api.invoker.HttpInvoker;
+import io.gravitee.gateway.reactive.api.policy.http.HttpPolicy;
+import io.gravitee.gateway.reactive.handlers.api.adapter.invoker.InvokerAdapter;
 import io.gravitee.policy.trafficshadowing.configuration.TrafficShadowingPolicyConfiguration;
 import io.gravitee.policy.trafficshadowing.invoker.ShadowInvoker;
+import io.gravitee.policy.trafficshadowing.v3.TrafficShadowingPolicyV3;
+import io.reactivex.rxjava3.core.Completable;
+import lombok.extern.slf4j.Slf4j;
 
-/**
- * @author David BRASSELY (david.brassely at graviteesource.com)
- * @author GraviteeSource Team
- */
-public class TrafficShadowingPolicy {
-
-    private final TrafficShadowingPolicyConfiguration configuration;
+@Slf4j
+public class TrafficShadowingPolicy extends TrafficShadowingPolicyV3 implements HttpPolicy {
 
     public TrafficShadowingPolicy(final TrafficShadowingPolicyConfiguration configuration) {
-        this.configuration = configuration;
+        super(configuration);
+    }
+
+    @Override
+    public String id() {
+        return "traffic-shadowing";
     }
 
-    @OnRequest
-    public void onRequest(ExecutionContext context, PolicyChain chain) {
-        // Override the invoker
-        Invoker defaultInvoker = (Invoker) context.getAttribute(ExecutionContext.ATTR_INVOKER);
-        context.setAttribute(ExecutionContext.ATTR_INVOKER, new ShadowInvoker(defaultInvoker, configuration));
-        chain.doNext(context.request(), context.response());
+    @Override
+    public Completable onRequest(HttpPlainExecutionContext ctx) {
+        return Completable.defer(() -> {
+            log.debug("Prepare Traffic Shadowing");
+
+            HttpInvoker defaultInvoker = ctx.getInternalAttribute(InternalContextAttributes.ATTR_INTERNAL_INVOKER);
+            if (defaultInvoker instanceof InvokerAdapter invokerAdapter) {
+                ctx.setInternalAttribute(
+                    InternalContextAttributes.ATTR_INTERNAL_INVOKER,
+                    new io.gravitee.policy.trafficshadowing.v3.invoker.ShadowInvoker(new InvokerAdapter(invokerAdapter), configuration)
+                );
+                return Completable.complete();
+            }
+
+            ctx.setInternalAttribute(InternalContextAttributes.ATTR_INTERNAL_INVOKER, new ShadowInvoker(defaultInvoker, configuration));
+            return Completable.complete();
```

Comment:
Suggestion: maybe litllely more easy to read?

```suggestion
            HttpInvoker defaultInvoker = ctx.getInternalAttribute(InternalContextAttributes.ATTR_INTERNAL_INVOKER);
            var shadowInvoker = defaultInvoker instanceof InvokerAdapter invokerAdapter
                ? new io.gravitee.policy.trafficshadowing.v3.invoker.ShadowInvoker(new InvokerAdapter(invokerAdapter), configuration)
                : new ShadowInvoker(defaultInvoker, configuration)
                );

            ctx.setInternalAttribute(InternalContextAttributes.ATTR_INTERNAL_INVOKER, shadowInvoker);
            return Completable.complete();
```